hellsemble/hellsemble.py

```diff
@@ -241,69 +248,175 @@ def __fit_estimators_sequential(
         # Creating copies of data for collecting fitting history
         X_fit, y_fit = X_train, y_train
         X_val, y_val = X_validation, y_validation
-        while (
-            self.estimator_generator.has_next()
-            and not self.__fitting_stop_condition(validation_fitting_history)
-            and len(np.unique(y_fit)) > 1
-        ):
-            # Generate next iterator
-            estimator = self.estimator_generator.fit_next_estimator(X_fit, y_fit)
-            self.estimators.append(estimator)
+        # Ensure numpy arrays for safe integer indexing (mypy fix)
+        X_train = np.asarray(X_train)
+        y_train = np.asarray(y_train)
+        X_validation = np.asarray(X_validation)
+        y_validation = np.asarray(y_validation)
+        best_score = 0
+        fit_indices = failed_observations_idx_fit.copy()
+        val_indices = failed_observations_idx_val.copy()
 
-            # Make and evaluate predictions on training set
-            fit_predictions = self.prediction_generator.make_prediction_train(
-                estimator, X_fit
-            )
-            failed_observations_mask_fit = fit_predictions != y_fit
-            coverage_counts.append(X_fit.shape[0] - failed_observations_mask_fit.sum())
+        self.progressive_train_scores = []
+        self.progressive_val_scores = []
 
-            # Make and evaluate predictions on validation set
-            val_predictions = self.prediction_generator.make_prediction_train(
-                estimator, X_val
-            )
-            failed_observations_mask_val = val_predictions != y_val
-
-            # Create prediction history entry
-            failed_observations_idx_fit = failed_observations_idx_fit[
-                failed_observations_mask_fit
-            ]
-            train_fitting_history_entry = np.full((X_train.shape[0]), False)
-            train_fitting_history_entry[failed_observations_idx_fit] = True
-            train_fitting_history.append(train_fitting_history_entry)
-
-            failed_observations_idx_val = failed_observations_idx_val[
-                failed_observations_mask_val
-            ]
-            validation_fitting_history_entry = np.full((X_validation.shape[0]), False)
-            validation_fitting_history_entry[failed_observations_idx_val] = True
-            validation_fitting_history.append(validation_fitting_history_entry)
-
-            # Update fitting and validation data
-            X_fit, y_fit = (
-                X_fit[failed_observations_mask_fit],
-                y_fit[failed_observations_mask_fit],
-            )
+        while not self.__fitting_stop_condition(validation_fitting_history):
+            best_model = None
+            best_ensemble_score = best_score
+            self.estimator_generator.reset_generator()
```

Comment:
I dont't get the point of this, shouldn't it be only in greedy approach? Afaik this line force in each iteration start from the very beggining so the same model is evaluate everytime?

Reply:
or maybe should it be outside this loop?